test/common/config/metadata_test.cc

```diff
@@ -49,6 +49,57 @@ TEST(MetadataTest, MetadataValuePath) {
             ProtobufWkt::Value::KindCase::KIND_NOT_SET);
 }
 
+// Test case for typed metadata: original Any data in metadata field
+TEST(MetadataTest, OriginalTypedMetadataValue) {
+  // create an any data and set its type_url and value
+  ProtobufWkt::Any any;
+  any.set_type_url("type.googleapis.com/foo");
+  any.set_value("bar");
+
+  // Put the data into metadata typed_filter_metadata field with the key : baz.
+  envoy::config::core::v3::Metadata metadata;
+  (*(metadata.mutable_typed_filter_metadata()))["baz"] = any;
+
+  // Verify the data is setup correctly in the metadata
+  EXPECT_EQ(metadata.typed_filter_metadata().find("test"), metadata.typed_filter_metadata().end());
+  EXPECT_NE(metadata.typed_filter_metadata().find("baz"), metadata.typed_filter_metadata().end());
+  EXPECT_EQ((*(metadata.mutable_typed_filter_metadata()))["baz"].type_url(),
+            "type.googleapis.com/foo");
+  EXPECT_EQ((*(metadata.mutable_typed_filter_metadata()))["baz"].value(), "bar");
+}
+
+// Test case for typed metadata: Pack the Any data in metadata field
+TEST(MetadataTest, PackedTypedMetadataValue) {
+  // create an any_source data and set its type_url and value
+  ProtobufWkt::Any any_source;
+  any_source.set_type_url("type.googleapis.com/foo");
+  any_source.set_value("bar");
+
+  // Pack this any_source data into metadata typed_filter_metadata
+  // field with the key : baz.
+  envoy::config::core::v3::Metadata metadata;
+  (*(metadata.mutable_typed_filter_metadata()))["baz"].PackFrom(any_source);
+
+  // Verify the data is setup correctly in the metadata
+  EXPECT_EQ(metadata.typed_filter_metadata().find("test"), metadata.typed_filter_metadata().end());
+  EXPECT_NE(metadata.typed_filter_metadata().find("baz"), metadata.typed_filter_metadata().end());
+  EXPECT_EQ((*(metadata.mutable_typed_filter_metadata()))["baz"].ShortDebugString(),
+            "[type.googleapis.com/google.protobuf.Any] { type_url: \"type.googleapis.com/foo\" "
+            "value: \"bar\" }");
+  EXPECT_EQ((*(metadata.mutable_typed_filter_metadata()))["baz"].type_url(),
+            "type.googleapis.com/google.protobuf.Any");
+  EXPECT_EQ((*(metadata.mutable_typed_filter_metadata()))["baz"].value(),
```

Comment:
Same as above, this seems somewhat fragile.

Reply:
Yeah, Let me remove this.